pkg/machine/cleanup.go

```diff
@@ -0,0 +1,66 @@
+package machine
+
+import (
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+
+	"github.com/sirupsen/logrus"
+)
+
+type CleanupCallback struct {
+	Funcs []func() error
+	mu    sync.Mutex
+}
+
+func (c *CleanupCallback) Clean(err *error) {
+	// do not remove created files if the init is successful
+	if *err == nil {
+		return
+	}
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	for _, cleanfunc := range c.Funcs {
+		if err := cleanfunc(); err != nil {
+			logrus.Error(err)
+		}
+	}
+}
+
+func (c *CleanupCallback) CleanOnSignal(err *error) {
+	ch := make(chan os.Signal, 1)
+	signal.Notify(ch, os.Interrupt, syscall.SIGTERM)
+
+	_, ok := <-ch
+	if !ok {
+		return
+	}
+	// if err exists, then we let the deferred clean function handle the cleanup
+	// this prevents a race where both defer and cleanonsignal tries to handle cleanups
+	if *err != nil {
+		return
+	}
+
+	c.mu.Lock()
+	for _, cleanfunc := range c.Funcs {
+		if err := cleanfunc(); err != nil {
+			logrus.Error(err)
+		}
+	}
+	c.mu.Unlock()
+	os.Exit(1)
+}
+
+func InitCleanup() CleanupCallback {
```

Comment:
Just a minor note  that technically you don't need an InitCleanup method here, since your append approach works with nil, and it fits go's zero-value-is-useful philosophy. Although maybe you wanted one anyway for future expansion, and thats great too.